Feat metadata

apachano opened the pull request:
Resolves: #121 
Authors:
@apachano 

## Summary of changes
- Implement a metadata storage system
- Integrate metadata storage system into blocks
- Record rotation on block placement in metadata on block
- Rotation is enabled during block-type registration
  
## Caveats
- Rotation does not render

## On approval
Merge

beeperdeeper089 commented on Phoenix/Common/Include/Common/Metadata.hpp:
```diff
@@ -0,0 +1,80 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/**
```
Do we want to bother standardising this? A few of our files have this but not that many.

apachano replied:
I've been trying to, it does make the doxygen output nicer but its not super critical so I'm not concerned about pushing it if that makes sense?

beeperdeeper089 replied:
Yeah sounds alright, I'll start telling Resharper to out that in my files too.

beeperdeeper089 commented on Phoenix/Common/Include/Common/Metadata.hpp:
```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <Common/Utility/Serializer.hpp>
+#include <any>
```
Put a newline here just for consistency between our includes and standard ones (since that's what we're doing in most of our other code).

beeperdeeper089 commented on Phoenix/Common/Include/Common/Metadata.hpp:
```diff
@@ -0,0 +1,80 @@
+	{
+	public:
+		/**
+		 * @brief sets or inserts metadata
```
Being a bit pedantic (and I'm also a victim of this) but capitals and full stops.

beeperdeeper089 commented on Phoenix/Common/Include/Common/Metadata.hpp:
```diff
@@ -0,0 +1,80 @@
+		 * @return false if the data already exists with a different data type
+		 */
+		bool set(const std::string& key, const std::any& existing);
+		/**
```
Newline between functions.

beeperdeeper089 commented on Phoenix/Common/Include/Common/Metadata.hpp:
```diff
@@ -0,0 +1,80 @@
+		 * @return A pointer to the data
+		 */
+		const std::any* get(const std::string& key) const;
+		/**
```
Newline.

beeperdeeper089 commented on Phoenix/Common/Include/Common/Metadata.hpp:
```diff
@@ -0,0 +1,80 @@
+		 */
+		void erase(const std::string& key) { m_data.erase(key); };
+
+		size_t size() { return m_data.size(); };
```
Use ``std::size_t``

beeperdeeper089 commented on Phoenix/Common/Include/Common/Metadata.hpp:
```diff
@@ -0,0 +1,80 @@
+		 * @brief Erases metadata
+		 * @param key The key associated with the metadata
+		 */
+		void erase(const std::string& key) { m_data.erase(key); };
```
What happens if the key doesn't exist? (I think it just doesn't do anything, but would just like to double check)

beeperdeeper089 commented on Phoenix/Common/Source/Metadata.cpp:
```diff
@@ -0,0 +1,125 @@
+		ser << data.first;
+		if (data.second.type() == typeid(int))
+		{
+			ser << std::string("int") << std::any_cast<int>(data.second);
```
1. Do you have to construct as ``std::string`` - if so that is yet another shortcoming of the serializer -_-. If that needs fixing I can shove in a patch for that.

2. I would argue against putting in the full data type, just putting ``i, f, s, v (int, float, string, vector)`` would be a better idea.

apachano replied:
You do have to construct as a string, if you don't then it will save as a cstring and in-turn will need retrieved as a cstring. Originally I put just a single character with the exception of a vec which may need expanded in the future to support additional types. 

We can convert this down to a single char - we just need to make a decision on how we would label a veci.

apachano replied:
I converted to a single char, we can figure out handling multiple vec types later

beeperdeeper089 commented on Phoenix/Common/Source/Metadata.cpp:
```diff
@@ -0,0 +1,125 @@
+		}
+		else
+		{
+			LOG_FATAL("Metadata")
```
We don't want to be able to get to a stage where this should ever happen - if there is an invalid type we've really fucked up somewhere but it shouldn't have to affect the general end user - so maybe this should be a debug message and it's just ignored otherwise?

apachano replied:
I totally spaced it in implementation but I feel we should keep fatal but kill the program. Leaving this unhanded will lead to corruption.

beeperdeeper089 replied:
Hmm, I'm not too sure, doing this leaves us in a spot of having many exit points through out the program (we definitely need to cut down on them), but I feel like metadata not being stored is not an issue enough that we should quit?

I'm not too sure, but my current argument is against exiting since I would assume it can't result in corruption or anything since it's just not outputting a piece of data.

apachano replied:
Well its not just not outputting data, its not serializing a value that we will attempt to read later. If it doesn't segfault on read we will get a chunk full of OOB/ random blocks instead of what we would want. At the very least we should abort the save to avoid corruption but since this will never happen -> going any further than throwing an exception is a bit overkill.

beeperdeeper089 commented on Phoenix/Common/Include/Common/Voxels/Chunk.hpp:
```diff
@@ -105,19 +106,38 @@ namespace phx::voxels
+		 * @param newBlock The block that now exists at this location.
 		 */
-		void setBlockAt(math::vec3 position, BlockType* newBlock);
+		void setBlockAt(math::vec3 position, Block newBlock);
```
``const math::vec3&``

beeperdeeper089 commented on Phoenix/Common/Include/Common/Voxels/Chunk.hpp:
```diff
@@ -105,19 +106,38 @@ namespace phx::voxels
+		 * @return false if the data already exists with a different data type
+		 * OR if the supplied position is out of bounds for the chunk.
+		 */
+		bool setMetadataAt(phx::math::vec3 position, const std::string& key,
```
Const ref.

beeperdeeper089 commented on Phoenix/Common/Include/Common/Voxels/Map.hpp:
```diff
@@ -77,7 +77,7 @@ namespace phx::voxels
 		    math::vec3 position);
 		BlockType* getBlockAt(math::vec3 position);
-		void       setBlockAt(math::vec3 pos, BlockType* block);
+		void       setBlockAt(math::vec3 pos, const Block& block);
```
Const ref on vec3.

beeperdeeper089 reviewed (changes requested):
Some basic changes, however a concern at the base level:

  - If we're supporting limited types, like ints, floats, vec3 (which I would assume would make sense - bare in mind I haven't put much thought into this past now) - does std::any really make sense or should we use something like std::variant which supports limited types and makes it easier for us to track/label types within it?

apachano commented:
@beeperdeeper089 I think I addressed everything, I also finished up everything I'm prepared to finish re rotation so there are a few changes I added making rotation optional.

beeperdeeper089 reviewed (commented):
Replied to a comment.

beeperdeeper089 commented on Phoenix/Client/Include/Client/Voxels/BlockRegistry.hpp:
```diff
@@ -129,16 +129,28 @@ namespace phx::client
+				    }
+
+				    sol::optional<bool> rotV = luaBlock["roth"];
+				    if (rotH)
```
I'm assuming you meant ``if (rotV)``

beeperdeeper089 commented on Phoenix/Client/Include/Client/Voxels/BlockRegistry.hpp:
```diff
@@ -129,16 +129,28 @@ namespace phx::client
+				    sol::optional<bool> rotV = luaBlock["roth"];
+				    if (rotH)
+				    {
+					    block.rotH = *rotV;
```
I'm assuming ``block.rotV``

beeperdeeper089 reviewed (changes requested):
You still have a few comments left to address and a few new ones.

apachano commented:
@beeperdeeper089 I think this is ready

tobyplowy commented:
Builds and runs. Can't find any issues.

beeperdeeper089 commented:
Looks good, just still the use of ``std::size_t`` in Metadata.hpp:69.

I'll approve and merge once that's in.

apachano commented:
@beeperdeeper089 Fixed the size_t
